### PR TITLE
Make sure published messages are acknowledged for play mode

### PR DESCRIPTION
**Pull request by Barry-Xu-2018**
Address #571

After test with the latest code for rolling (use fastdds), #571 (follow steps to reproduce in #571) cannot reproduce.   
Note that synchronous is set as default for fastdds  recently(https://github.com/ros2/rmw_fastrtps/commit/e7c749ae167b7b08cc904737c6eac16e51399b02).  But even if I use async mode, #571 also cannot reproduce.

### CODE REVIEW

**Comment by Barry-Xu-2018** [comment]
The failure on ament_uncrustify is related to file `mock_storage.hpp`, which isn't changed in this PR.

```
--- rosbag2_compression/test/rosbag2_compression/mock_storage.hpp
+++ rosbag2_compression/test/rosbag2_compression/mock_storage.hpp.uncrustify
@@ -43 +43 @@
-    void(const std::vector<std::shared_ptr<const rosbag2_storage::SerializedBagMessage>> &));
+    void(const std::vector<std::shared_ptr<const rosbag2_storage::SerializedBagMessage>>&));

Code style divergence in file 'rosbag2_cpp/test/rosbag2_cpp/mock_storage.hpp':

--- rosbag2_cpp/test/rosbag2_cpp/mock_storage.hpp
+++ rosbag2_cpp/test/rosbag2_cpp/mock_storage.hpp.uncrustify
@@ -44 +44 @@
-    void(const std::vector<std::shared_ptr<const rosbag2_storage::SerializedBagMessage>> &));
+    void(const std::vector<std::shared_ptr<const rosbag2_storage::SerializedBagMessage>>&));

2 files with code style divergence
Error: The process '/usr/bin/bash' failed with exit code 1
```

**Comment by Barry-Xu-2018** [comment]
@fujitatomoya @emersonknapp   
Please help to review.

**Comment by fujitatomoya** [comment]
@Barry-Xu-2018 will do in a couple of days.

**Review by jhdcs — APPROVED** [review]
Looks good to me. I only saw a couple comments that might use a bit of tweaking, though I don't think it's enough to hold up the review over.

Since you requested @fujitatomoya and @emersonknapp as reviewers, specifically, I'll wait on their judgement rather than attempt merging for now.

**Comment by fujitatomoya** [comment]
@jhdcs since i do not have any authorization on this repository, you can just go ahead instead of my comments if you are good to go. (i will do the review as community reviewer :smiley: )

File: rosbag2_transport/src/rosbag2_transport/player.cpp [hunk 1]
```diff
@@ -229,6 +229,30 @@ void Player::play()
...
+    std::chrono::milliseconds timeout(play_options_.wait_acked_timeout);
+    if (timeout == std::chrono::milliseconds(0)) {
+      timeout = std::chrono::milliseconds(-1);
+    }
+    for (auto pub : publishers_) {
+      try {
+        if (!pub.second->wait_for_all_acked(timeout)) {
+          RCLCPP_ERROR(
+            get_logger(),
+            "Failed to wait all published messages acknowledged for topic %s",
```
Comment (fujitatomoya) [inline rosbag2_transport/src/rosbag2_transport/player.cpp:244]: this happens only when `timeout`, so how about the following? I think timeout is different from failure or exception.

```suggestion
            "Timeout to wait all published messages acknowledged for topic %s",
```

File: rosbag2_transport/src/rosbag2_transport/player.cpp [hunk 2]
```diff
@@ -483,6 +507,14 @@ void Player::prepare_publishers()
       publishers_.insert(
         std::make_pair(
           topic.name, create_generic_publisher(topic.name, topic.type, topic_qos)));
+      if (play_options_.wait_acked_timeout >= 0 &&
+        topic_qos.reliability() == rclcpp::ReliabilityPolicy::BestEffort)
+      {
+        RCLCPP_WARN(
+          get_logger(),
+          "--wait-for-all-acked is invaild for topic '%s' since reliability of QOS is BestEffort.",
+          topic.name.c_str());
```
[the remark this answers is not in the document]
Comment (Barry-Xu-2018) [reply rosbag2_transport/src/rosbag2_transport/player.cpp:516]: > if 100 of 200 topics come true with this if statement, there will be 100 warning lines here. 

Yes. It is poor user experience.  
But I think topic information is useful for user.   
How about change as below: (only print one warning for all invalid topics)   
```
-wait-for-all-acked is invalid for below topics with reliability of QOS is BestEffort
xxx, xxx, xxx, xxx, xxx 
```

File: ros2bag/ros2bag/verb/play.py [hunk 3]
```diff
@@ -35,6 +36,13 @@ def positive_float(arg: str) -> float:
     return value
 
 
+def not_negative_int(arg: str) -> int:
+    value = int(arg)
+    if value < 0:
+        raise ArgumentTypeError(f'Value {value} is less than zero.')
+    return value
+
+
```
[the remark this answers is not in the document]
Comment (Barry-Xu-2018) [reply ros2bag/ros2bag/verb/play.py:45]: According to my understanding, the input parameter of check function is fixed (only `arg: str`, and this parameter is passed by argparse module).   
We cannot use this parameter to specify what to do, such as check `positive` or `not_negative`.   
So I have no idea how to use one common function to implement `check_positive_float` and `not_negative_int`.

[the remark this answers is not in the document]
Comment (fujitatomoya) [reply ros2bag/ros2bag/verb/play.py:45]: i was not clear on this, i was just saying to move this function where `check_positive_float` is.

File: ros2bag/ros2bag/verb/play.py [hunk 4]
```diff
@@ -93,6 +101,17 @@ def add_arguments(self, parser, cli_name):  # noqa: D102
         parser.add_argument(
             '--start-offset', type=check_positive_float, default=0.0,
             help='Start the playback player this many seconds into the bag file.')
+        parser.add_argument(
+            '--wait-for-all-acked', type=not_negative_int, default=0,
```
[the remark this answers is not in the document]
Comment (Barry-Xu-2018) [reply ros2bag/ros2bag/verb/play.py:105]: Yeah.   
I think this option should be disabled by default, since most cases don't need this feature.   
What do you think ?

File: ros2bag/ros2bag/verb/play.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Barry-Xu-2018) [reply ros2bag/ros2bag/verb/play.py:45]: >  i was just saying to move this function where check_positive_float is.

Get it.

File: ros2bag/ros2bag/verb/play.py [hunk 5]
```diff
@@ -93,6 +94,17 @@ def add_arguments(self, parser, cli_name):  # noqa: D102
         parser.add_argument(
             '--start-offset', type=check_positive_float, default=0.0,
             help='Start the playback player this many seconds into the bag file.')
+        parser.add_argument(
+            '--wait-for-all-acked', type=check_not_negative_int, default=-1,
+            help='Wait until all published messages are acknowledged by all subscribers or until '
+                 'the timeout elapses in millisecond before play is terminated. '
+                 'Especially for the case of sending message with big size in a short time. '
```
Comment (jhdcs) [inline ros2bag/ros2bag/verb/play.py:101]: I feel like this flows a little better, though I tend to be a bit wordy.

```suggestion
                 'Especially useful when sending many messages with large sizes in a short timeframe.'
```

File: rosbag2_transport/src/rosbag2_transport/player.cpp [hunk 6]
```diff
@@ -229,6 +229,30 @@ void Player::play()
...
+    std::chrono::milliseconds timeout(play_options_.wait_acked_timeout);
+    if (timeout == std::chrono::milliseconds(0)) {
+      timeout = std::chrono::milliseconds(-1);
+    }
+    for (auto pub : publishers_) {
+      try {
+        if (!pub.second->wait_for_all_acked(timeout)) {
+          RCLCPP_ERROR(
+            get_logger(),
+            "Timeout to wait all published messages acknowledged for topic %s",
```
Comment (jhdcs) [inline rosbag2_transport/src/rosbag2_transport/player.cpp:244]: I feel like this reads a little better.

```suggestion
            "Timed out while waiting for all published messages to be acknowledged for topic %s",
```

File: rosbag2_transport/src/rosbag2_transport/player.cpp [hunk 7]
```diff
@@ -229,6 +229,30 @@ void Player::play()
...
+            pub.first.c_str());
+        }
+      } catch (std::exception & e) {
+        RCLCPP_ERROR(
+          get_logger(),
+          "Failed to wait all published messages acknowledged for topic %s : %s",
```
Comment (jhdcs) [inline rosbag2_transport/src/rosbag2_transport/player.cpp:250]: Another small wordage suggestion.

```suggestion
          "Exception occurred while waiting for all published messages to be acknowledged for topic %s : %s",
```

File: rosbag2_transport/src/rosbag2_transport/player.cpp [hunk 8]
```diff
@@ -491,6 +521,18 @@ void Player::prepare_publishers()
...
+
+  if (!topic_without_support_acked.empty()) {
+    // remove the last ", "
+    topic_without_support_acked.erase(
+      topic_without_support_acked.end() - 2,
+      topic_without_support_acked.end());
+
+    RCLCPP_WARN(
+      get_logger(),
+      "--wait-for-all-acked is invaild for below topics since reliability of QOS is BestEffort.\n"
```
Comment (jhdcs) [inline rosbag2_transport/src/rosbag2_transport/player.cpp:533]: Very minor change here.

```suggestion
      "--wait-for-all-acked is invaild for the below topics since reliability of QOS is BestEffort.\n"
```

**Review by jhdcs — APPROVED** [review]
Did another review pass, since several changes were made. Again, just had some comment/documentation suggestions.

File: rosbag2_transport/src/rosbag2_transport/player.cpp (hunk 8 above)
Comment (Barry-Xu-2018) [reply rosbag2_transport/src/rosbag2_transport/player.cpp]: Thanks for your comments.   
Please help to check again @jhdcs

**Review by jhdcs — APPROVED** [review]
Looks good to me!

**Comment by clalancette** [comment]
> Looks good to me!

Did CI from https://ci.ros2.org get run on this PR?

**Comment by jhdcs** [comment]
Oh crud, I thought it had!

Crud, let me see how to undo that...

**Comment by jhdcs** [comment]
Not sure I did that right...

Sorry, first time merging a PR...

**Comment by clalancette** [comment]
Here's CI after the merge:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/16048/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/10703/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/16372/)

**Comment by fujitatomoya** [comment]
@clalancette thanks for checking on this one. after all, everything looks okay!
